source/common/stats/utility.h

```diff
@@ -12,6 +13,32 @@
 namespace Envoy {
 namespace Stats {
 
+/**
+ * Represents a dynamically created stat name token based on absl::string_view.
+ * This class wrapper is used in the 'Element' variant so that call-sites
+ * can express explicit intent to create dynamic stat names, which are more
+ * expensive than symbolic stat names. We use dynamic stat names only for
+ * building stats based on names discovered in the line of a request.
+ */
+class DynamicName : public absl::string_view {
+public:
+  // This is intentionally left as an implicit conversion from string_view to
+  // make call-sites easier to read, e.g.
+  //    Utility::counterFromElements(*scope, {DynamicName("a"), DynamicName("b")});
+  using absl::string_view::string_view;
+};
+
+/**
+ * Holds either a symbolic StatName or a dynamic string, for the purpose of
+ * composing a vector to pass to Utility::counterFromElements, etc. This is
+ * a programming convenience to create joined stat names. It is easier to
+ * call the above helpers than to use SymbolTable::join(), because the helpers
+ * hide the memory management of the joined storage, and they allow easier
+ * co-mingling of symbolic and dynamic stat-name components.
+ */
+using Element = absl::variant<StatName, DynamicName>;
+using ElementVec = std::vector<Element>;
```

Comment:
I would imagine that the number of elements here in general will be pretty short. Did you consider absl::InlinedVector here with some small default size? You could likely avoid heap allocations in many places, especially when using RVO (from some of the examples I see in this PR).

Reply:
good idea; done. I think I considered that and forgot why I backed off. Seems to work fine, at least for the few tests I tried.

I think we should also do that for StatNameVec but that's a bigger change for reasons. Dont want to pollute this PR.

Reply:
Yeah that's fine. Perhaps in that change can we try to cover both of these in a micro-benchmark? Might be worth it to try to tease out some of the improvements here.